### ISSUE TITLE
launch Visual Studio Installer to install the C++ tools

### ISSUE THREAD

**Issue by ctaggart** [rust-lang/rustup#1462]
`rust-init.exe` does a good job at detecting for missing C++ tools. It would be even better if it detected an existing Visual Studio installation to prompt with something like: `Would you like to launch Visual Studio Installer to install the C++ tools?` and then do so.

The scenario is that I have Visual Studio 2017 installed and I want to install Rust. Currently you get this prompt https://github.com/rust-lang-nursery/rustup.rs/issues/1114#issuecomment-405064524 . It would be amazing if `rust-init.exe` launched the Visual Studio Installer to install the missing dependencies. Here is a PowerShell script that I ran as admin to get them installed:

``` ps1
# start the Visual Studio Installer to install Rust dependencies
# current component names are:
#   VC++ 2017 version 15.7 v14.14 latest v141 tools
#   Windows 10 SDK (10.0.17134.0)

# get the latest Visual Studio installation path
$installationPath = & "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe" -latest -property installationPath

# launch the installer (must be admin)
& "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vs_installershell.exe" modify `
    --installPath $installationPath `
    --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 `
    --add Microsoft.VisualStudio.Component.Windows10SDK
```

I took notes in https://github.com/ctaggart/rustup.rs/issues/1 while making this.

**Comment by Diggsey** [comment]
I would be happy to accept a PR for this, although we'd need a fallback for detecting visual studio: I have VS2017 installed and `vswhere` is not present.

**Comment by ctaggart** [comment]
To include the preview Visual Studio version, you must add `-prerelease` to the first command. This just worked for me on a computer with only Visual Studio Community 2019 Preview.
``` ps1
$installationPath = & "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe" -latest -property installationPath -prerelease
& "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vs_installershell.exe" modify `
    --installPath $installationPath `
    --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 `
    --add Microsoft.VisualStudio.Component.Windows10SDK
```

**Comment by ctaggart** [comment]
An alternative to this is to add a `.vsconfig` file (based on the Terminal [.vsconfig](https://github.com/microsoft/terminal/blob/master/.vsconfig)) to your project and open Visual Studio there.

``` json
{
  "version": "1.0",
  "components": [
    "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
    "Microsoft.VisualStudio.Component.Windows10SDK"
  ]
}

**Comment by kinnison** [comment]
Is there a safe and consistent way we can launch this from `rustup-init.exe` ?

**Comment by cataggar** [comment]
It [turns out](https://github.com/ctaggart/rustup.rs/issues/1#issuecomment-408621180) that the `Microsoft.VisualStudio.Component.Windows10SDK` component is actually the "Windows Universal C Runtime`. You can see that at:
https://docs.microsoft.com/visualstudio/install/workload-component-id-vs-community

I exported my current config and I do not have the component installed. I think the required components are now:

``` json
{
  "version": "1.0",
  "components": [
    "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
    "Microsoft.VisualStudio.Component.VC.Redist.14.Latest",
  ]
}
```
If I save this to a `visual-cpp-build-tools.vsconfig`, I can open Visual Studio Installer and `import configuration` to install any missing components:

![image](https://user-images.githubusercontent.com/87583576/127405868-8501b3db-6840-4b2d-a125-18826d05fa8d.png)

You can see it select those two `Individual components`, and I already have them installed:

![image](https://user-images.githubusercontent.com/87583576/127408627-9d19374e-5c6f-4bde-945d-33767ac4f74e.png)

It would be great if someone could add a screenshot without them installed yet.

**Comment by rami3l** [comment]
Fixed by #2954.
